modeshape-jcr/src/main/java/org/modeshape/jcr/value/binary/S3BinaryStore.java

```diff
@@ -268,45 +269,61 @@ public InputStream getInputStream(BinaryKey key) throws BinaryStoreException {
-
-        // There is no capacity in S3 to query on object properties. This must be done
-        // by straight iteration, so may take a very long time for large data sets.
+    /**
+     * Sets a tag on a S3 object, potentially overwriting the existing value.
+     */
+    private void setS3ObjectTag(String objectKey, String tagKey, String tagValue) throws BinaryStoreException {
         try {
-            for(BinaryKey key : getAllBinaryKeys()) {
-                ObjectMetadata metadata = s3Client.getObjectMetadata(bucketName, key.toString());
-                String unused = metadata.getUserMetadata().get(UNUSED_KEY);
-                if (null != unused && unused.equals(String.valueOf(true))) {
-                    Date lastMod = metadata.getLastModified();
-                    if (lastMod.before(deadline)) {
-                        try {
-                            s3Client.deleteObject(bucketName, key.toString());
-                        } catch (AmazonClientException e) {
-                            Logger log = Logger.getLogger(getClass());
-                            log.warn(e, JcrI18n.unableToDeleteTemporaryFile, e.getMessage());
-                        }
-                    }
-                } // Assumes that if no value is set, content is used
+            GetObjectTaggingRequest getTaggingRequest = new GetObjectTaggingRequest(bucketName, objectKey);
+            GetObjectTaggingResult getTaggingResult = s3Client.getObjectTagging(getTaggingRequest);
+
+            List<Tag> initialTagSet = getTaggingResult.getTagSet();
+            List<Tag> mergedTagSet = mergeS3TagSet(initialTagSet, new Tag(tagKey, tagValue));
+
+            if (initialTagSet.size() == mergedTagSet.size() && initialTagSet.containsAll(mergedTagSet)) {
+                return;
             }
+
+            SetObjectTaggingRequest setObjectTaggingRequest = new SetObjectTaggingRequest(bucketName, objectKey,
+                    new ObjectTagging(mergedTagSet));
+
+            s3Client.setObjectTagging(setObjectTaggingRequest);
         } catch (AmazonClientException e) {
             throw new BinaryStoreException(e);
         }
     }
 
+    /**
+     * Merges a new tag into an existing list of tags.
+     * It will be either appended to the list or overwrite the value of an existing tag with the same key.
+     */
+    private List<Tag> mergeS3TagSet(List<Tag> initialTags, Tag changeTag) {
+        Map<String, String> mergedTags =  initialTags.stream().collect(Collectors.toMap(Tag::getKey, Tag::getValue));
+        mergedTags.put(changeTag.getKey(), changeTag.getValue());
+        return mergedTags.entrySet().stream().map(
+                entry -> new Tag(entry.getKey(), entry.getValue())).collect(Collectors.toList());
+    }
+
+    @Override
+    public void removeValuesUnusedLongerThan(long minimumAge, TimeUnit timeUnit) throws BinaryStoreException {
+        // XXX
+        // What should this method do?
```

Comment:
I was thinking, maybe the original functionality and code should be left in place, as default behavior.
This would be disabled and turned into a no-op if the user explicitly configures the S3 binary store with an option like, say, `out-of-band-deletion = true`.

Reply:
@dalbani can you help to educate the community of nascent ModeShape archaeologists ;) on the old code vs. this patch and the implications of your suggestion here? @bbranan any input from you as the original author of the S3 binary store would also be appreciated.

Reply:
I find that reading MODE-2698 adds a lot of context. For my part, since it appears that using the Amazon lifecycle would require additional configuration on the part of the user, I agree it would be best to leave the original code in place and implement this as a (recommended) optional feature.

Reply:
This approach has the benefit of essentially removing the need for Modeshape to directly manage the removal of any unused files. If a bucket is large, this could provide a significant performance benefit. What is lost is control over purge timing, which could be a significant problem for some use cases. Regardless, this method will almost certainly no longer abide by the contract defined in the Interface (i.e. it will no longer remove any unused values).

I believe there is significant value in this new approach, but sticking with the original behavior as the default seems like the safest model to me, as it is closest to the expected behavior. 

I don't see where the bucket lifecycle rule is put in place. I assume that is yet to be added.

Reply:
My S3-fu is as yet nonexistent. Do APIs exist whereby ModeShape could configure that rule rather than requiring the user to install it?

Reply:
Yes, there are API calls in S3 that would allow for creating the bucket rule, so that could be done within Modeshape. The user would only need to make sure that the permissions (IAM) policy associated with the AWS credentials they supply are sufficient to allow bucket rules to be added.

Reply:
@bbranan Bill, thanks for your response. It appears that @dalbani has lost interest; could you please weigh in on my question at https://github.com/ModeShape/modeshape/pull/1660/files#r233589792 ?

Reply:
If I remember correctly (boy, a year and a half ago already), my idea was to accomplish 2 things regarding the S3 storage backend:
1. use tags instead of copy objects around (due to object immutability) to keep costs down
2. use S3 lifecycle API which happens to support tags, in order to free ModeShape from a possibly tedious task in case of a repository with lots of unused objects.

To be honest, I've moved off S3 quite quickly and hosted my data in Google Cloud Storage. I have written my own GCS-based storage implementation for ModeShape, largely inspired by the S3 one.
I found GCS more appropriate for my needs (no eventual consistency, object properties can be updated, etc).

I happen to be in the process of migrating a ModeShape repository with 3+ TB / 7+ million binaries to Apache Oak, which seems to be the only active JCR implementation.
ModeShape has been quite stable over the last year and a half, although I ran into some limitations with querying in particular (e.g. JOINs) and indexes (take ages to rebuild).
But, yeah, ModeShape is unfortunately dead, even though I never saw any official announcement by the way, right?
Oh, I also wrote a PostgreSQL module to be able to store BSON data from ModeShape as native JSON binary in the database.
The data is not a "black box" anymore and it can be nicely compressed by PostgreSQL!

Reply:
@blafond has taken over as the lead of ModeShape, and I have been trying to assist him in getting the outstanding PRs merged to `master`. This is the last one (modulo one which I am considering as a duplicate). It is my understanding that we are going to try to make a 5.5 release soon, and to keep the project alive going forward. Any information or code you can provide to improve ModeShape, even as you depart :| , would be, as far as I'm concerned, appreciated.

Matt